### DISCUSSION TITLE
Dependency Injection KernelBuilder

### DISCUSSION THREAD

**Discussion by CKThompson00** [microsoft/semantic-kernel#6471] [Q&A] [answered]
I have a customer who had a question around DI/IoC with regards to KernelBuilder.  They have dependencies they inject in the startup of the application for things that managed RAG and other non-kernel related work.  ILogger is a good example here.  They need to inject that in the startup as well as in the KernelBuilder.  Is this the right approach?  Does the Kernel run it's own container separate from the rest of the application?  If so, is what they are doing the right way (inject in startup and in the KernelBuilder)?

**Comment by stephentoub** [comment, marked as answer]
KernelBuilder is typically just for getting started in an environment that lacks its own DI container. The more typical real-app usage is in ASP.NET or Maui or Azure Functions or some other environment where there is a DI container, in which case you just integrate with that (e.g. the `AddKernel` extension for `IServiceCollection`). See https://youtu.be/jrNfKeGSuCg?t=179, and in particular starting at 15:56 (https://youtu.be/jrNfKeGSuCg?t=957). Kernel itself just wraps an arbitrary IServiceProvider, which can come from anywhere. If you use Kernel.CreateBuilder, that's creating a new standalone service collection that's then passed off to the built Kernel when you call Build. If you instead just use `AddKernel` (which is doing little more than `AddTransient<Kernel>()`), when you request a kernel from that DI container, it's just creating a Kernel that wraps that overall shared IServiceProvider, so all other services in that provider, including ILogger-related stuff, will all be available.

**Reply by CKThompson00** [reply]
Thanks @stephentoub.  This helps a lot and addressed the customers concern.  Looks like Sophia marked as answered.

**Comment by sabbadino** [comment]
but what if I want to build different kernels , inject them in asp.net as keyed services ? 
Why don't we have a AddKernel(string kernelname) ?

**Reply by nguyenlamlll** [reply]
Yeah, it's missing that thing. And I feel AddKernel does not allow us to configure anything. Or it just lacks serious documentation.
Even in this devblog, they still suggest AddKeyedTransient: https://devblogs.microsoft.com/semantic-kernel/using-semantic-kernel-with-dependency-injection/
And doc: https://learn.microsoft.com/en-us/semantic-kernel/concepts/kernel?pivots=programming-language-csharp#using-dependency-injection
